### PR TITLE
mod_push fixes

### PR DESCRIPTION
**Pull request by astro**
Hi

These are just three little fixes. Justifications are in the commit messages.

### CODE REVIEW

File: apps/ejabberd/src/mod_push_service_mongoosepush.erl [hunk 1]
```diff
@@ -44,7 +44,8 @@
 %%--------------------------------------------------------------------
 
 -spec start(Host :: ejabberd:server(), Opts :: list()) -> any().
-start(Host, Opts) ->
+start(ServerHost, Opts) ->
```
Comment (gadgetci) [inline apps/ejabberd/src/mod_push_service_mongoosepush.erl]: According to **Compiler**:
> Warning: variable 'ServerHost' is unused

File: apps/ejabberd/src/mod_push_service_mongoosepush.erl [hunk 2]
```diff
@@ -44,7 +44,8 @@
 %%--------------------------------------------------------------------
 
 -spec start(Host :: ejabberd:server(), Opts :: list()) -> any().
-start(Host, Opts) ->
+start(_ServerHost, Opts) ->
+    Host = ?MYNAME,
```
Comment (rslota) [inline apps/ejabberd/src/mod_push_service_mongoosepush.erl]: This change may lead to module crashes when the server is configured with several virtual-hosts, because the module will start multiple times with the same `Host`. Why this is an issue for you? Sure, the current code will start multiple VHosts and the pubsub will use only one of them, but "it should work". The correct fix here would be not to use `?MYNAME ` in `node_push`. I don't really remember why I did that, but it may be because it was not trivial to get the host in that context.

Anyway, if you let me know what the real issue is, we may find better solution to this problem.

File: apps/ejabberd/src/node_push.erl [hunk 3]
```diff
@@ -207,20 +208,16 @@ is_allowed_to_publish(PublishModel, Affiliation) ->
 parse_form(undefined) ->
     #{};
 parse_form(Form) ->
-    IsForm = ?NS_XDATA == exml_query:attr(Form, <<"xmlns">>),
-    IsSubmit = <<"submit">> == exml_query:attr(Form, <<"type">>),
```
Comment (rslota) [inline apps/ejabberd/src/node_push.erl]: `type` is mandatory here. XEP-0357 specifies that this push node shall get `publish-options` form. The XEP-0060 (Publish-Subscribe) says that form inside of `publish-options` has to be the one defined in XEP-0004 (Data Forms). Finally XEP-0004 defines that `type` field of `x` element is `required` and has to be set to one of the following: `submit`, `cancel`, `result`, `form`. For reference see: [https://xmpp.org/extensions/xep-0004.html#schema](https://xmpp.org/extensions/xep-0004.html#schema)

Comment (astro) [inline apps/ejabberd/src/node_push.erl]: I was referring to [XEP-0357](https://xmpp.org/extensions/xep-0357.html#publishing). Multiple implementations could already exist that do not fit this requirement.

[Please be liberal in what you accept from others.](https://en.wikipedia.org/wiki/Robustness_principle)

Comment (michalwski) [inline apps/ejabberd/src/node_push.erl]: Hi @astro,

I understand this `type` check was an issue for you. Could you give us more details? What implementation do you have in mind?

File: apps/ejabberd/src/mod_push_service_mongoosepush.erl (hunk 2 above)
Comment (michalwski) [inline apps/ejabberd/src/mod_push_service_mongoosepush.erl]: I agree with @rslota, there is many installations where we need multiple XMPP host served by the same MongooseIM node.

The root cause is most probably in other place: https://github.com/esl/MongooseIM/pull/1222/files#diff-64730cfd8dcd5857c353e7ed3413f59dL108 The implementation now is bound to the main domain anyway, which is not a perfect solution. I somehow missed that while doing code review. The `Host` parameter which should be used is Publisher's host if I understand it correctly. @rslota, what's your opinion on that?

@astro could you give us more details what was the main issue for you here?

Comment (rslota) [inline apps/ejabberd/src/mod_push_service_mongoosepush.erl]: @michalwski Yes, it should. The issue was that `Host` is not available in the `node_push'a` callbacks. I must have forgotten to to change `mod_pubsub` to include this variable in the `publish` callback. I'm gonna take care of that.

File: apps/ejabberd/src/node_push.erl (hunk 3 above)
Comment (rslota) [inline apps/ejabberd/src/node_push.erl]: @astro @michalwski I didn't see that there is a mistake in examples from [XEP-0357](https://xmpp.org/extensions/xep-0357.html#publishing). Because of that we may skip this check but not this way in my opinion. My proposition is that we leave this check active, but we can default to `submit` form `type` in case this attribute is missing. This way, the current implementations that don't send this attribute will work and the code will reject wrong form types like `cancel`. What do you think about it @michalwski ?

Comment (michalwski) [inline apps/ejabberd/src/node_push.erl]: I agree with you @rslota

File: apps/ejabberd/src/mod_push_service_mongoosepush.erl (hunk 2 above)
Comment (rslota) [inline apps/ejabberd/src/mod_push_service_mongoosepush.erl]: @michalwski Fixed in #1233 , so Now the change from this PR is not needed / applicable. Therefore, @astro, could you please revert this change after #1233 is merged?

File: apps/ejabberd/src/node_push.erl (hunk 3 above)
Comment (rslota) [inline apps/ejabberd/src/node_push.erl]: @michalwski Fixed in #1234 :) @astro , could you please revert this change after #1234 is merged?

Comment (astro) [inline apps/ejabberd/src/node_push.erl]: Great, I'm happy with this solution and will remove the commit from this PR.

Comment (michalwski) [inline apps/ejabberd/src/node_push.erl]: #1234 is merged, waiting for travis to merge #1233 as well

**Comment by astro** [comment]
Ok, removed the two obsolete commits. Just the approved one is left now.

I'm also happy about e7919ddf11ff76aa4d0046a0a8ea24e3544554f9 which duplicates a change I would have pushed as well. :)

**Comment by michalwski** [comment]
Hi @astro,

Any chance you rebase your branch or solve conflicts soon? I'd like to merge your patch tomorrow.

**Comment by michalwski** [comment]
All right, we cherry-picked your changes in #1245

**Comment by michalwski** [comment]
Thanks for the patch @astro
